### ISSUE TITLE
[RFC] Validation of catalog definition via pipeline/github action (script)

### ISSUE THREAD

**Issue by jvilimek** [backstage/backstage#4834]
**Status:** Open for comments

## Need

**TL/DR**: looking for a way, how to validate backstage.yaml definition from a script.

We have partially generated and partially manually created definitions for backstage. We have backend services index, frontend components (spa, npm packages), manual entries.. all in git repositories. This is done as catalog-xxx/index.yaml of **kind=Location** with list of component definitions. Some of them are in the same (central) repository, many of them are references to respective service/spa/npm...repository. 
catalog-xxx/index.yaml is located in backstage definition and loaded at start + refreshed on a regular base:
```yaml
catalog:
  rules:
    - allow: [Component, API, Group, User, Template, Location]
  locations:
    # Backstage catalog index
    - type: url
      target: https://.../shared-repo/_git/backstage-catalog-components?path=%2Fcatalog-backend%2Findex.yaml
    - type: url
      target: https://.../shared-repo/_git/backstage-catalog-components?path=%2Fcatalog-frontend%2Findex.yaml
    - type: url
      target: https://.../shared-repo/_git/backstage-catalog-components?path=%2Fcatalog-manual-entries%2Findex.yaml
```

Example of one of the index file:
```
apiVersion: backstage.io/v1alpha1
kind: Location
metadata:
  name: backend-catalog-index
  description: A collection of all backend components
spec:
  type: url
  targets:
    - https://.../shared-repo/_git/backstage-catalog-components?path=%2Fcatalog-backend%2Fauto-generated%2FService01.yaml
    - https://.../service02-repo/_git/backend-service?path=%2Fbackstage.yaml #resource: Service02
```

Definition of Service01 is autogenerated, definition of Service02 is referenced to its repository. 

Now when the definition is changed manually by the team of Service02 and it is broken, the whole index import might be broken so we are looking for a way, how to prevent this..

Our need is to have a way (e.g. prepare some pipeline/action) to **validate during the pull request**, that the Service02 backstage.yaml config is valid.

## Proposal

Prepare a suggested implementation of Backstage including pipeline/github action templates to validat the backstage catalog YAML file definitions.

Can we e.g. call some backstage backend API to validate the url of current service? Should this be synchronous call? Asynchronous? Proposal is to build a working pipeline sample, full process how to develop/adjust backstage.yaml definitions in service repositories.

## Alternatives

Complete script solution without the call of the backstage API ...just read the definition, validate we have syntactically correct YAML, that we can read all locations, that the tags are in correct form, to validate just the schema...

## Enhancements

We can validate the schema, that it is OK. But what would be super uber cool is also having some semantic rules to check. E.g. that we have always sonarQube project definition for backend services (generates a warning), specific labels filled in, links (e.g. for DEV, STG, LIVE environments), that the name is unique or that the API components ends via "-api" etc.

## Risks

the API can be misused. We need to somehow authenticate/authorize the call. We need to have managed identity/API key passed so we can execute the API validation.

**Comment by dhenneke** [comment]
There is a `dryRun` flag for the `POST /locations` endpoint that is also used by the catalog-import plugin which works synchronous:
https://github.com/backstage/backstage/blob/98387ac6b5f348054088424aa2ebb7fe6685438d/plugins/catalog-backend/src/service/router.ts#L116-L120

It just tries to add it to the database but doesn't commit the transaction so nothing is actually written. But as you said, authentication could be a problem.

More detailed checks or backend-independent steps still need more thought. However, I'm not sure if it would really work externally, since you won't be able to verify that the locations you link to are actually accessible from the backend in the location processor.

**Comment by Fox32** [comment]
Related #4336

**Comment by dtuite** [comment]
@jvilimek  We've created a validator library based on requests from Roadie customers. There's a plain NPM package, GitHub action and Circle CI Orb available. You can find it here: https://github.com/RoadieHQ/backstage-entity-validator

It has the following features so far:

1. YAML syntax validation
1. Checking that required Backstage entity properties are in place
2. Checking that annotations are spelled correctly
3. Checking that annotation values are correctly prefixed with "url" where appropriate
4. Checking that URLs are valid

LMK if this is useful please. We want to test it out a bit first and see if it's useful but we're totally open to contributing parts of it upstream.

**Comment by Rugvip** [comment]
I'd say the https://github.com/RoadieHQ/backstage-entity-validator covers an MVP. Alright to close this and create more specific issues for things we want to add?

**Comment by jvilimek** [comment]
Nice! That's basicaly what I was looking for! Will try asap.

**Comment by jvilimek** [comment]
> @jvilimek We've created a validator library based on requests from Roadie customers. There's a plain NPM package, GitHub action and Circle CI Orb available. You can find it here: https://github.com/RoadieHQ/backstage-entity-validator
> ...
> LMK if this is useful please. We want to test it out a bit first and see if it's useful but we're totally open to contributing parts of it upstream.

Thanks a lot. I somehow missed this message. Will test it (in Azure DevOps) asap. Maybe I can help creating the extension but I guess simple use via npm would be just ok.
